### PR TITLE
SEAB-5951: Notebook "launch to GitHub Codespace" support

### PR DESCRIPTION
**Pull request by svonworl**
**Description**
This PR modifies the webservice to read the devcontainer file(s) that are usually present in a repo that's intended to be launched in a GitHub Codespace.

When a repo is opened in a codespace, GitHub could inspect the repo to infer the type/contents (hey, there's an `ipynb` file, so this must be a notebook, fire up the notebook support), but it doesn't.  Instead, it reads the devcontainer file(s) from the repo, which specify the [optional] image that will run in the Codespace container, any vscode extensions that will be started, command to run, files to open, etc.   The devcontainer spec:

https://containers.dev/

Specifically, this PR adds a new file type `DOCKSTORE_NOTEBOOK_DEVCONTAINER`, and modifies the Jupyter language handler to find and read devcontainer files as SourceFiles of that type.

The intention is that the UI will use this information to advise the user at launch time.  For example, when a notebook repo is launched to a codespace but there's not devcontainer in the repo, the notebook will not automatically load.  If we can determine this at "launch to codespace" time, we can inform the user and tell them what to do ("click on X/Y/Z when the interface loads", "add this suggested devcontainer", etc)...

Possibly, in the future, the launch process for non-GitHub platforms could also inspect the devcontainer to determine which image to use, etc.
 
**Review Instructions**
Clone/create a repo that contains a devcontainer file, register it in qa, and confirm that the devcontainer `SourceFile` appears and has the correct type when you retrieve it via the appropriate endpoint. 

**Issue**
https://ucsc-cgl.atlassian.net/browse/SEAB-5951

**Security and Privacy**
No concerns.

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] Follow the existing JPA patterns for queries, using named parameters, to avoid SQL injection
- [x] If you are changing dependencies, check the Snyk status check or the dashboard to ensure you are not introducing new high/critical vulnerabilities
- [x] Assume that inputs to the API can be malicious, and sanitize and/or check for Denial of Service type values, e.g., massive sizes
- [x] Do not serve user-uploaded binary images through the Dockstore API
- [x] Ensure that endpoints that only allow privileged access enforce that with the `@RolesAllowed` annotation
- [x] Do not create cookies, although this may change in the future
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead.

### CODE REVIEW

File: dockstore-integration-testing/src/test/java/io/dockstore/client/cli/ExtendedMetricsTRSOpenApiIT.java [hunk 1]
```diff
@@ -86,7 +86,6 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
-
```
Comment (coverbeck) [inline dockstore-integration-testing/src/test/java/io/dockstore/client/cli/ExtendedMetricsTRSOpenApiIT.java:89]: Heh, remember our merge issue yesterday? :) But there's a ticket on this so we can deal with it later.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/languages/JupyterHandler.java [hunk 2]
```diff
@@ -134,29 +142,57 @@ public Map<String, SourceFile> processImports(String repositoryId, String conten
...
+        } else if (rootNames.contains(DOT_DEVCONTAINER_DIR)) {
+            String path = "/" + DOT_DEVCONTAINER_DIR;
+            Set<String> names = listFiles.apply(path);
+
+            if (names.contains(DEVCONTAINER_JSON)) {
+                // Read /.devcontainer/devcontainer.json
+                fileReader.accept(path + "/" + DEVCONTAINER_JSON, DOCKSTORE_NOTEBOOK_DEVCONTAINER);
+
+            } else {
+                // Scan for and read all files with path /.devcontainer/<folder>/devcontainer.json
```
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/languages/JupyterHandler.java:186]: What are the implications of multiple devcontainer.json files? Does GitHub Codespace automatically pick one when you run the notebook, or prompt you?

If it automatically picks one, should we only index the one that it picks, assuming it's deterministic?

Comment (svonworl) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/languages/JupyterHandler.java:186]: Here's an example of the link we'll probably load in the new window, pointing at the branch with the two devcontainers:
https://github.com/codespaces/new?hide_repo_select=true&ref=dotdir-folder-devcontainers&repo=dockstore-testing/simple-notebook&skip_quickstart=true&machine=basicLinux32gb&geo=UsWest
You can select the devcontainer configuration.  Interesting thing is, on the branches where there's one devcontainer, selecting "Default project configuration" will use the devcontainer, but on the branch with two devcontainers, selecting "Default project configuration" will use _neither_.  This raises the possibility that maybe you can have another default devcontainer on that branch, too.  I will do an experiment and see if that's true.

Comment (svonworl) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/languages/JupyterHandler.java:186]: Experimental results: Github reads and gives you the option to launch with all of the devcontainers with spec-conformant paths.  It does not register devcontainers from other paths in the tree.  So, we should probably grab all of the former.

**Comment by svonworl** [comment]
Hello y'all.  I discovered that GitHub will read all of the devcontainers, and [at least in some cases] allow the user to choose which environment [devcontainer] they want to use in the codespace.  So, I changed the webservice to read all of the devcontainers that are at the prescribed paths within the repo.  Re-requesting review...

**Review by kathy-t — APPROVED** [review]

**Review by coverbeck — APPROVED** [review]

**Pull request merged**